### DISCUSSION TITLE
Integrate this project in already existing gqlgen project

### DISCUSSION THREAD

**Discussion by frederikhors** [nautilus/gateway#145] [Q&A] [answered]
_I'm sorry for this question, I'm still learning GraphQL and Golang world._

I'm using `gqlgen` (https://github.com/99designs/gqlgen) to generate my own GraphQL server.

I need to "merge" other private/custom schemas to my own.

But I don't want to start two different services (my own gqlgen server and nautilus/gateway) to do so.

Is there a way to "integrate", in the same Golang project, gqlgen generated server and this amazing gateway?

In this way I can reduce a lot the network requests: there would be only those for "external" schemas.

What do you think of my idea?

**Comment by frederikhors** [comment]
I already have a:

```go
var gql = handler.GraphQL(graphql.NewExecutableSchema(graphql.Config{Resolvers: &graphql.Resolver{}}), config.GraphqlHandlerOptions)
```

It would be amazing to use that `graphql.NewExecutableSchema` coming from `gqlgen`.

**Comment by edwinmo-splunk** [comment, marked as answer]
Hey @frederikhors! We actually had this exact use case and came up with a way to not have to host two separate services and still utilize the Nautilus gateway. We accomplished this by writing our own [Queryer](https://github.com/nautilus/graphql/blob/master/queryer.go#L48) that will resolve the query using the gqlgen executable schema. It looks roughly like this:

```
func (self *customQueryer) Query(ctx context.Context, input *graphqlGW.QueryInput, receiver interface{}) error {
	// ignore the input.QueryDocument as it seems that Nautilus is not constructing it correctly
	// instead, rely on the input.Query like how gqlgen wrote their handler
	doc, gqlErr := parser.ParseQuery(&ast.Source{Input: input.Query})
	if gqlErr != nil {
		return gqlErr
	}

	// more validation stuff; this is REQUIRED and not just a validation step
	// it changes the `*ast.Schema` that is required for gqlgen to execute it
	if listErr := validator.Validate(self.executableSchema.Schema(), doc); len(listErr) != 0 {
		return fmt.Errorf("Error validating query with schema %+v", listErr)
	}

	requestContext := graphql.NewRequestContext(doc, input.Query, input.Variables)
	newCtx := graphql.WithRequestContext(ctx, requestContext)
	operationDefinition := doc.Operations.ForName(input.OperationName)

	var response *graphql.Response
	if operationDefinition.Operation == ast.Mutation {
		response = self.executableSchema.Mutation(newCtx, operationDefinition)
	} else {
		response = self.executableSchema.Query(newCtx, operationDefinition)
	}

	responseBytes, err := response.Data.MarshalJSON()
	if err != nil {
		return err
	}

	err = json.Unmarshal(responseBytes, receiver)
	if err != nil {
		return err
	}

	return nil
}
```

Here, `self.executableSchema` is the result from calling `graphql.NewExecutableSchema` coming from `gqlgen`.

You'll need to provide a [QueryerFactory](https://github.com/nautilus/gateway/blob/master/plan.go#L64).

I hope this was helpful.
